### ISSUE TITLE
Line number is 0 in dict returned from getbufinfo()

### ISSUE THREAD

**Issue by rhysd** [vim/vim#7999]
**Describe the bug**
`getbufinfo()` returns buffer information as dict and it includes `lnum` key which indicates current line number. `getbufinfo()` incorrectly returns `0` for `lnum` when current buffer is not the buffer.

**To Reproduce**
Detailed steps to reproduce the behavior:
1. Run `vim --clean`
2. `:new` (it opens 2 buffers: bufnr=1, bufnr=2)
3. `:new` again (it opens 1 buffer: bufnr=3)
4. `:echo getbufnr(2)[0].lnum`

**Expected behavior**

It echoes `1`

**Screenshots**

<img width="461" alt="スクリーンショット 2021-03-24 1 10 34" src="https://user-images.githubusercontent.com/823277/112179143-c327ef00-8c3d-11eb-95d5-c20d364515d6.png">

It echoes `0`

**Environment (please complete the following information):**
 - Vim version: 8.2.2646
 - OS: macOS 10.15
 - Terminal: iTerm2

**Additional context**

`lnum` is still `0` even if the buffer (bufnr=2) contains text of multiple lines in the case.

When moving cursor to the window which opens the buffer (bufnr=2), `echo getbufinfo(2)[0].lnum` echoes `1` as intended.

**Comment by lacygoill** [comment]
Duplicate of https://github.com/vim/vim/issues/6612.

---

See [`:h getbufinfo()`](https://vimhelp.org/eval.txt.html#getbufinfo()):

> lnum          Line number used for the buffer when
>               opened in the current window.

Notice the "when opened in the current window" part.

**Comment by chrisbra** [comment]
closing then

**Comment by rhysd** [comment]
Oh, thank you for the point. I didn't notice that. So the value is useless when it is not opened in the current window...

**Comment by lacygoill** [comment]
@rhysd If your Vim includes the patch [8.2.1727](https://github.com/vim/vim/releases/tag/v8.2.1727), and if your buffer is currently displayed in a window, you can get the address of the line where the cursor was the last time that window was active, by passing its window number or window ID as an argument to `getcurpos()`:

    :echo getcurpos(2)[1]

**Comment by lacygoill** [comment]
And if your Vim includes the patch [8.1.1967](https://github.com/vim/vim/releases/tag/v8.1.1967), you can get the same information with `line()`:

    :echo line('.', win_getid(2))

**Comment by lacygoill** [comment]
I think this is a common mistake.  Maybe we should improve the documentation:
```diff
diff --git a/runtime/doc/eval.txt b/runtime/doc/eval.txt
index 86df7b570..6d3aac7e9 100644
--- a/runtime/doc/eval.txt
+++ b/runtime/doc/eval.txt
@@ -5097,8 +5097,16 @@ getbufinfo([{dict}])
 					last used.
 					{only with the |+viminfo| feature}
 			listed		TRUE if the buffer is listed.
+
 			lnum		Line number used for the buffer when
 					opened in the current window.
+					NOTE: Only useful if the buffer has
+					been displayed in the window in the
+					past.  If you want the line number
+					of the last known cursor position in
+					a given window, use |line()|: >
+						:echo line('.', {winid})
+<
 			linecount	Number of lines in the buffer (only
 					valid when loaded)
 			loaded		TRUE if the buffer is loaded.
```

**Comment by rhysd** [comment]
Yes, I could use `line()` instead. One trouble for me was that I needed to use different functions for supporting both Vim and Neovim since Neovim does not support the second parameter of `line()`. But it is not an issue of Vim.

> I think this is a common mistake. Maybe we should improve the documentation:

It looks useful. I needed to look around for alternative. With the note, I would be able to know the alternative instantly.
